lib/stages/visAccumulate.cpp

```diff
@@ -231,6 +231,8 @@ void visAccumulate::register_base_dataset_states(
     base_dataset_states.push_back(dm.create_state<inputState>(inputs).first);
     base_dataset_states.push_back(dm.create_state<prodState>(prods).first);
     base_dataset_states.push_back(dm.create_state<eigenvalueState>(0).first);
+    base_dataset_states.push_back(
```

Comment:
If you add it to the base dataset states I think it means that when you actually have a gated stream it will have a new gatingState in its ancestors as well as this one. That sort of suggests that the input has been gated twice (even though it hasn't), so I think you probably want to add the uniform gating state elsewhere.

Reply:
@jrs65 I'm unsure where to put it then. Should it be in the main loop:

```
    while (!stop_thread) {

        // Fetch a new frame and get its sequence id
        uint8_t* in_frame = wait_for_full_frame(in_buf, unique_name.c_str(), in_frame_id);
        if (in_frame == nullptr)
            break;

        // Check if dataset ID changed
        dset_id_t ds_id_in_new = get_dataset_id(in_buf, in_frame_id);
        if (!ds_id_in || ds_id_in_new != *ds_id_in) {
            ds_id_in = ds_id_in_new;

            // Register base dataset. If no dataset ID was was set in the incoming frame,
            // ds_id_in will be dset_id_t::null and thus cause a root dataset to
            // be registered.
            base_dataset_id = dm.add_dataset(base_dataset_states, *ds_id_in);
            DEBUG("Registered base dataset: {}", base_dataset_id)

            // Set the output dataset ID for the main visibility accumulation
            gated_datasets.at(0).output_dataset_id = base_dataset_id;
        }   

        int32_t* input = (int32_t*)in_frame;
        uint64_t frame_count = (get_fpga_seq_num(in_buf, in_frame_id) / samples_per_data_set);

        // Start and end times of this frame
        timespec t_s = ((chimeMetadata*)in_buf->metadata[in_frame_id]->metadata)->gps_time;
        timespec t_e = add_nsec(t_s, samples_per_data_set * tel.seq_length_nsec());
```

Reply:
Okay. Yeah, I think the reason this isn't super obvious is because of the bug that is being fixed in here.

I think in the if statement in the code above, when it detects that the dataset ID has changed, it needs to update all the gated_datasets with new dataset IDs.

Essentially the line starting with `gated_datasets.at(0).output_dataset_id ...` needs to get replaced with a loop over gated_datasets and each one needs to have its `output_dataset_id` replaced with the return value of `register_gate_dataset(spec)`.

That should take care of it all, and you can remove the gate dataset registration from the base dataset states.